MINIXCompat/MINIXCompat_Filesystem.c

```diff
@@ -570,4 +727,199 @@ minix_fd_t MINIXCompat_File_Access(const char *minix_path, minix_mode_t minix_mo
 }
 
 
+// MARK: - Directories
+
+static void MINIXCompat_Dir_CopyName(char *to, const char *from, uint16_t fromlen)
+{
+    assert(to != NULL);
+    assert(from != NULL);
+    assert(fromlen <= 14);
+
+    if (fromlen == 14) { // up to 14 characters without trailing NUL
+        strncpy(to, from, 13);
+        to[13] = from[13]; // no trailing NUL
+    } else {
+        strncpy(to, from, fromlen);
+    }
+}
+
+/*! Pre-cache a directory for reading. */
+static int16_t MINIXCompat_Dir_Precache(const char * _Nullable host_path, minix_fd_t minix_fd)
+{
+    // Open the directory for iteration.
+
+    DIR *dir = opendir(host_path);
+    if (dir == NULL) {
+        return -MINIXCompat_Errors_MINIXErrorForHostError(errno);
+    }
+
+    // Create a minix_dirent_t for every corresponding struct dirent, resizing our table as needed.
+
+    bool done_reading = false;
+    size_t entry_count = 0;
+    size_t dircache_count = 32; // start at 32 since most directories have fewer entries, and 32 are one MINIX block
+
+    minix_dirent_t *dircache = calloc(dircache_count, sizeof(minix_dirent_t));
+
+    do {
+        errno = 0;  // errno will be unchanged for end-of-directory
+        struct dirent *entry = readdir(dir);
+        if (entry == NULL) {
+            int host_errno = errno;
+            if (host_errno != 0) { // errno is unchanged for end-of-directory
+                return -MINIXCompat_Errors_MINIXErrorForHostError(host_errno);
+            } else {
+                // Read has finished successfully. Close the directory.
+                closedir(dir);
+                dir = NULL;
+            }
+            done_reading = true;
+        } else {
+            // Expand the cache if needed.
+
+            if (entry_count >= dircache_count) {
+                // Always increase size by one block's worth so it doesn't grow too quickly.
+                size_t new_dircache_count = (dircache_count + 32);
+                dircache = realloc(dircache, new_dircache_count * sizeof(minix_dirent_t));
+
+                // Zero-fill new entries.
+                memset(&dircache[dircache_count], 0, 32);
+
+                // Update number of available entries.
+                dircache_count = new_dircache_count;
+            }
+
+            // MINIX just wants inode and 14-character name.
+            // Note that a full 14-character name isn't zero-terminated, so use a special copy function.
+            // Also note that d_ino needs swapping in order to be passed back to MINIX, do that here since nothing host-side uses it.
+
+            dircache[entry_count].d_ino = htons(MINIXCompat_File_MINIXInodeForHostInode(entry->d_ino));
+            MINIXCompat_Dir_CopyName(dircache[entry_count].d_name, entry->d_name, entry->d_namlen);
```

Comment:
[POSIX dirent](https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/dirent.h.html) only contains `d_ino` and `d_name`, and at least glibc does not include `d_namlen`. Fortunately, it is redundant  since `d_name` will be a NUL-terminated string.